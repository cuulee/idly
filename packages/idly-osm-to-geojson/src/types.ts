import { Entity } from 'idly-common/lib/osm/structures';

export interface Derived {
  readonly entity: Entity;
  readonly parentWays: Set<string>;
  readonly parentRelations: Set<string>;
}
