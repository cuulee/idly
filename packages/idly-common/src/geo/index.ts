export * from '../geo/sphericalMercator';
export * from '../geo/bbox';
