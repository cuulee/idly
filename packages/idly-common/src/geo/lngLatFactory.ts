import { LngLat } from '../osm/structures';

export function lngLatFactory(
  obj: { lon: number; lat: number } | [number, number]
): LngLat {
  if (Array.isArray(obj)) {
    return { lon: obj[0], lat: obj[1] };
  }
  return { ...obj };
}
