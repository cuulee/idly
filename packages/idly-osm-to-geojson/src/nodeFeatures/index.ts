import {
  Feature,
  Point,
} from '@turf/helpers';
import { presetMatch } from 'idly-common/lib/geojson/presetMatch';
import { Node, OsmGeometry } from 'idly-common/lib/osm/structures';
import { Derived } from '../types';

export const DEFAULT_NODE_ICON = 'circle';

export function nodeFeatures(derived: Derived<Node>): Feature<Point> {
  const node = derived.entity;
  return {
    geometry: {
      coordinates: [node.loc.lon, node.loc.lat],
      type: 'Point',
    },
    id: node.id,
    properties: nodePropertiesGen(derived),
    type: 'Feature',
  };
}

export const nodePropertiesGen = (element: Derived<Node>) => {
  const size = element.parentWays.size;
  const geometry = size === 0 ? OsmGeometry.POINT : OsmGeometry.VERTEX;
  const match = presetMatch(element.entity.tags, geometry);
  return {
    '@idly-geometry': geometry,
    '@idly-icon': (match && match.icon) || DEFAULT_NODE_ICON,
    '@idly-intersection': size > 2,
    '@idly-name': element.entity.tags.name,
    id: element.entity.id,
  };
};
