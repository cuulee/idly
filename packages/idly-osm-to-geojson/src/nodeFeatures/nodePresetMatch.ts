import { presetMatch } from 'idly-common/lib/geojson/presetMatch';
import { Node, OsmGeometry } from 'idly-common/lib/osm/structures';
import { Derived } from '../types';

export function nodePresetMatch(element: Derived<Node>) {
  const size = element.parentWays.size;
  const geometry = size === 0 ? OsmGeometry.POINT : OsmGeometry.VERTEX;
  const match = presetMatch(element.entity.tags, geometry);

  return { match, geometry };
}
