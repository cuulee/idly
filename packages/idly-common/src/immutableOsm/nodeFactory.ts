import { $attributesGen } from './attributesGen';
import { $genLngLat } from './genLngLat';
import { $tagsFactory } from './tagsFactory';
import { $Node, $__Node } from '../immutableOsm';
import { genLngLat } from '../osm/genLngLat';
import { attributesGen } from '../osm/attributesGen';
import { Attributes, EntityId, EntityType, LngLat, Node, Tags } from '../osm';
import { recordify } from 'typed-immutable-record';

export function nodeFactory(n: Node): $Node {
  var $tags = $tagsFactory(n.tags);
  var $loc = $genLngLat(n.loc);
  var $attr = $attributesGen(n.attributes);
  return recordify<$__Node, $Node>({
    id: n.id,
    tags: $tags,
    type: EntityType.NODE,
    loc: $loc,
    attributes: $attr
  });
}
