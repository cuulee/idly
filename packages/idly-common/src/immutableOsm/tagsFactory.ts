import { Tags } from '../osm/structures';
import { $Tags } from '../immutableOsm/immutableOsm';
import { Map as $Map } from 'immutable';

export function $tagsFactory(t: Tags): $Tags {
  return $Map(t);
}
