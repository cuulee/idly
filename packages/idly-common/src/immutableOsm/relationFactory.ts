import { List as $List } from 'immutable';
import { $relationMemberGen } from './relationMemberGen';
import { $attributesGen } from './attributesGen';
import { $tagsFactory } from './tagsFactory';
import { $Relation, $__Relation } from '../immutableOsm';
import { attributesGen } from '../osm/attributesGen';
import {
  Attributes,
  EntityId,
  EntityType,
  Relation,
  RelationMember,
  Tags
} from '../osm';
import { recordify } from 'typed-immutable-record';

export function $relationFactory(r: Relation): $Relation {
  var $tags = $tagsFactory(r.tags);
  var $attr = $attributesGen(r.attributes);
  var $members = $List(r.members.map($relationMemberGen));

  return recordify<$__Relation, $Relation>({
    id: r.id,
    type: EntityType.RELATION,
    tags: $tags,
    attributes: $attr,
    members: $members
  });
}
