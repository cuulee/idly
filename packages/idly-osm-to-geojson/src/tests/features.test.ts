import { lngLatFactory } from 'idly-common/lib/geo/lngLatFactory';
import { nodeFactory } from 'idly-common/lib/osm/entityFactory/nodeFactory';
import { relationFactory } from 'idly-common/lib/osm/entityFactory/relationFactory';
import { tagsFactory } from 'idly-common/lib/osm/entityFactory/tagsFactory';

import { wayFactory } from 'idly-common/lib/osm/entityFactory/wayFactory';
import {
  Node,
  OsmGeometry,
  Relation,
  Way,
} from 'idly-common/lib/osm/structures';
import { nodeFeatures } from '../nodeFeatures';
import { Derived, DerivedTable } from '../types';
import {
  getCoordsFromTable,
  wayFeatures,
  wayToLineString,
} from '../wayFeatures';

describe('converts node to feat', () => {
  const n1: Derived<Node> = {
    entity: nodeFactory({ id: 'n1' }),
    parentRelations: new Set(),
    parentWays: new Set(['w1']),
  };

  const n11 = {
    entity: nodeFactory({
      id: 'n1#0',
      loc: lngLatFactory({ lon: 15, lat: 10 }),
      tags: tagsFactory({ k: 'k' }),
    }),
    parentRelations: new Set(),
    parentWays: new Set(),
  };
  it('should behave...', () => {
    expect(nodeFeatures(n1).properties).toMatchObject({
      '@idly-geometry': OsmGeometry.VERTEX,
    });

    expect(nodeFeatures(n11).properties).toMatchObject({
      '@idly-geometry': OsmGeometry.POINT,
    });

    expect(nodeFeatures(n1)).toMatchSnapshot();
    expect(nodeFeatures(n11)).toMatchSnapshot();
  });
});

describe('way.test', () => {
  describe('wayToLineString', () => {
    it('should throw error when no geometry provided', () => {
      expect(() => wayToLineString(undefined as any, [[1, 2]], {})).toThrow(); // ts:disable-line
      expect(() => wayToLineString(OsmGeometry.POINT, [[1, 2]], {})).toThrow();
    });

    it('should not work when correct coords provided with wrong geometry', () => {
      expect(() =>
        wayToLineString(OsmGeometry.POINT, [[1, 2], [2, 3], [3, 0], [1, 2]], {})
      ).toThrow();
    });
    it('should throw error when wrong coords provided', () => {
      expect(() =>
        wayToLineString(OsmGeometry.AREA, [[1, 2], [2, 1]], {})
      ).toThrow();
    });
    it('should work when correct coords provided', () => {
      expect(() =>
        wayToLineString(OsmGeometry.AREA, [[1, 2], [2, 3], [3, 0], [1, 2]], {})
      ).not.toThrow();
    });
    it('should behave...', () => {
      expect(
        wayToLineString(OsmGeometry.AREA, [[1, 2], [2, 3], [3, 0], [1, 2]], {})
      ).toMatchSnapshot();
    });
  });

  describe('get Nodes', () => {
    it('should get nodes correctly', () => {
      const node: Derived<Node> = {
        entity: nodeFactory({ id: 'n1', loc: lngLatFactory([1, 2]) }),
        parentRelations: new Set(),
        parentWays: new Set(['w1']),
      };

      const g: DerivedTable = new Map();

      g.set(node.entity.id, node);

      expect(getCoordsFromTable(g, ['n1'])).toEqual([[1, 2]]);
    });

    it('should throw error if nodes not found', () => {
      const g: DerivedTable = new Map();

      expect(() => getCoordsFromTable(g, ['n2'])).toThrow();
    });

    it('should work when multiple nodes are there in graph', () => {
      const node1: Derived<Node> = {
        entity: nodeFactory({ id: 'n1', loc: lngLatFactory([1, 2]) }),
        parentRelations: new Set(),
        parentWays: new Set(['w1']),
      };

      const node2: Derived<Node> = {
        entity: nodeFactory({ id: 'n2', loc: lngLatFactory([1, 4]) }),
        parentRelations: new Set(),
        parentWays: new Set([]),
      };

      const g = new Map();
      g.set(node1.entity.id, node1);
      g.set(node2.entity.id, node2);

      expect(getCoordsFromTable(g, ['n2'])).toEqual([[1, 4]]);
      expect(getCoordsFromTable(g, ['n1', 'n2'])).toEqual([[1, 2], [1, 4]]);
    });
  });

  describe('wayFeatures', () => {
    const n1: Derived<Node> = {
      entity: nodeFactory({ id: 'n1', loc: lngLatFactory([1, 2]) }),
      parentRelations: new Set(),
      parentWays: new Set(['w1']),
    };
    const n2: Derived<Node> = {
      entity: nodeFactory({ id: 'n2', loc: lngLatFactory([3, 4]) }),
      parentRelations: new Set(),
      parentWays: new Set(['w1']),
    };

    const n3: Derived<Node> = {
      entity: nodeFactory({ id: 'n3', loc: lngLatFactory([1, 4]) }),
      parentRelations: new Set(['r1']),
      parentWays: new Set([]),
    };

    const w1: Derived<Way> = {
      entity: wayFactory({
        id: 'w1',
        nodes: ['n1', 'n2'],
        tags: tagsFactory({ highway: 'residential' }),
      }),
      parentRelations: new Set(),
      parentWays: new Set(),
    };

    const r1: Derived<Relation> = {
      entity: relationFactory({
        id: 'r1',
        members: [{ id: 'n1', ref: 'n1' }, { id: 'n2', ref: 'n2' }],
        tags: {},
      }),
      parentRelations: new Set(),
      parentWays: new Set(),
    };

    const g: DerivedTable = new Map();

    [n1, n2, n3, w1, r1].forEach(e => {
      g.set(e.entity.id, e);
    });

    it('should behave...', () => {
      const result = wayFeatures(w1, g);
      expect(result).toMatchSnapshot();
    });

    it('sets up existing props', () => {
      const result = wayFeatures(w1, g);
      expect(result.properties).toMatchObject({
        '@idly-geometry': 'line',
        id: 'w1',
      });
    });
  });
});
