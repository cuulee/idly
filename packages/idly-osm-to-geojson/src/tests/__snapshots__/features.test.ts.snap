// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`converts node to feat should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      0,
      0,
    ],
    "type": "Point",
  },
  "id": "n1",
  "properties": Object {
    "@idly-geometry": "vertex",
    "@idly-icon": "circle",
    "@idly-intersection": false,
    "@idly-name": undefined,
    "@idly-preset-name": "Other",
    "id": "n1",
  },
  "type": "Feature",
}
`;

exports[`converts node to feat should behave... 2`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      15,
      10,
    ],
    "type": "Point",
  },
  "id": "n1#0",
  "properties": Object {
    "@idly-geometry": "point",
    "@idly-icon": "circle",
    "@idly-intersection": false,
    "@idly-name": undefined,
    "@idly-preset-name": "Point",
    "id": "n1#0",
  },
  "type": "Feature",
}
`;

exports[`way.test wayFeatures should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        1,
        2,
      ],
      Array [
        3,
        4,
      ],
    ],
    "type": "LineString",
  },
  "id": "w1",
  "properties": Object {
    "@idly-geometry": "line",
    "@idly-icon": "highway-residential",
    "@idly-isOneway": false,
    "@idly-name": undefined,
    "@idly-preset-name": "Residential Road",
    "@idly-tag-highway": "tag-highway-residential",
    "id": "w1",
  },
  "type": "Feature",
}
`;

exports[`way.test wayToLineString should behave... 1`] = `
Object {
  "geometry": Object {
    "coordinates": Array [
      Array [
        Array [
          1,
          2,
        ],
        Array [
          2,
          3,
        ],
        Array [
          3,
          0,
        ],
        Array [
          1,
          2,
        ],
      ],
    ],
    "type": "Polygon",
  },
  "properties": Object {},
  "type": "Feature",
}
`;
